compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java

```diff
@@ -255,10 +255,42 @@ private BPackageSymbol definePackage(DataInputStream dataInStream, int pkgCpInde
         // Define service declarations
         defineSymbols(dataInStream, rethrow(this::defineServiceDeclarations));
 
+        populateReferencedFunctions();
+
         this.typeReader = null;
         return this.env.pkgSymbol;
     }
 
+    private void populateReferencedFunctions() {
+        for (BStructureTypeSymbol structureTypeSymbol : this.structureTypes) {
+            if (structureTypeSymbol instanceof BObjectTypeSymbol) {
+                BObjectType objectType = (BObjectType) structureTypeSymbol.type;
+                for (BType typeRef : objectType.typeInclusions) {
+                    if (typeRef.tsymbol == null || typeRef.tsymbol.kind != SymbolKind.OBJECT) {
+                        continue;
+                    }
+
+                    List<BAttachedFunction> attachedFunctions = ((BObjectTypeSymbol) typeRef.tsymbol).attachedFuncs;
+                    for (BAttachedFunction function : attachedFunctions) {
+                        String referencedFuncName = function.funcName.value;
+                        Name funcName = names.fromString(
+                                Symbols.getAttachedFuncSymbolName(structureTypeSymbol.name.value, referencedFuncName));
+                        Scope.ScopeEntry matchingObjFuncSym = this.env.pkgSymbol.scope.lookup(funcName);
+                        if (matchingObjFuncSym != null) {
+                            continue;
+                        }
+                        if (Symbols.isPrivate(function.symbol)) {
+                            // we should not copy private functions.
+                            continue;
+                        }
```

Comment:
Shouldn't this be the first thing we do within the for loop?

Reply:
Updated in 1b6b21ff9a8ad716b114fd76ad04f19dc70a57e8